components/backup/src/service.rs

```diff
@@ -2,52 +2,57 @@
 
 use std::sync::atomic::*;
 
-use engine_traits::{KvEngine, RaftEngine};
 use futures::{channel::mpsc, FutureExt, SinkExt, StreamExt, TryFutureExt};
 use grpcio::{self, *};
 use kvproto::brpb::*;
-use raftstore::store::{
-    fsm::store::RaftRouter,
-    msg::{PeerMsg, SignificantMsg},
-};
-use tikv_util::{error, info, worker::*};
+use raftstore::store::snapshot_backup::{SnapshotBrHandle, UnimplementedHandle};
+use tikv_util::{error, info, warn, worker::*};
 
 use super::Task;
+use crate::disk_snap::{self, StreamHandleLoop};
 
 /// Service handles the RPC messages for the `Backup` service.
-#[derive(Clone)]
-pub struct Service<EK: KvEngine, ER: RaftEngine> {
+pub struct Service<H: SnapshotBrHandle> {
     scheduler: Scheduler<Task>,
-    router: Option<RaftRouter<EK, ER>>,
+    snap_br_env: disk_snap::Env<H>,
 }
 
-impl<EK, ER> Service<EK, ER>
-where
-    EK: KvEngine,
-    ER: RaftEngine,
-{
+impl<H: SnapshotBrHandle> Clone for Service<H> {
```

Comment:
Rustc will complain this:

```
This will send a message to every region in the store. I guess that will be fast if there isn't a region blocking.

error[E0277]: the trait bound `service::Service<UnimplementedHandle>: Clone` is not satisfied
   --> components/backup/src/service.rs:179:76
    |
179 |             ServerBuilder::new(env.clone()).register_service(create_backup(backup_service));
    |                                                              ------------- ^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `service::Service<UnimplementedHandle>`
    |                                                              |
    |                                                              required by a bound introduced by this call
    |
rustcClick for full compiler diagnostic
service.rs(179, 62): required by a bound introduced by this call
service.rs(15, 10): required for `service::Service<UnimplementedHandle>` to implement `Clone`
brpb_grpc.rs(144, 41): required by a bound in `kvproto::brpb::create_backup`
service.rs(179, 76): consider borrowing here: `&`
```

Reply:
make `SnapshotBrHandle: Sync + Send + Clone`

Reply:
The `SnapshotBrHandle` doesn't need to be `Clone` because it is `Arc`ed in the field. But somehow I agree that allow the `Handle` to implement `Clone` will be better.